std/stdio.d

```diff
@@ -4522,70 +4521,70 @@ Initialize with a message and an error code.
     }
 }
 
-extern(C) void std_stdio_static_this()
+// Undocummented but public because std* handles are aliasing it
+ref File makeGlobal(alias handle)()
 {
-    static import core.stdc.stdio;
-    //Bind stdin, stdout, stderr
-    __gshared File.Impl stdinImpl;
-    stdinImpl.handle = core.stdc.stdio.stdin;
-    .stdin._p = &stdinImpl;
-    // stdout
-    __gshared File.Impl stdoutImpl;
-    stdoutImpl.handle = core.stdc.stdio.stdout;
-    .stdout._p = &stdoutImpl;
-    // stderr
-    __gshared File.Impl stderrImpl;
-    stderrImpl.handle = core.stdc.stdio.stderr;
-    .stderr._p = &stderrImpl;
+    static __gshared File.Impl impl;
+    static __gshared File result;
+    static shared bool initialized;
+    if (!initialized)
+    {
+        // Use double-checking because initOnce is inefficient.
```

Comment:
> Careful with the word observed. Non-atomic reads – for the sake of the argument, let's use the C++11 memory model and consider the read from initialized relaxed instead of plainly being a data race – don't come with any sequential consistency guarantees, and so the execution order constraints you can infer from not taking the if are minimal.

Reads from `shared` integrals should be atomic, according to TDPL. They should not interleave with writes. I.e. there should not be a data race there. 

> The longer, hopefully more enlightening version in very much of a Socratic spirit (not that there would much of a basis for discussion if we don't assume, say, C++11 semantics for D):
> 
>     initOnce can't optimize it.
> 
> What about literally just moving the if inside initOnce?

We can't. It's of arbitrary type, might not even be convertible to bool, or `true` might not actually be the expected value.

>     It has to take the (potentially) fenced approach in order to make sure it does indeed store the initializer only once
> 
> That's already guaranteed by the check that makes the "doubled" part of the double-checked locking pattern, isn't it?

You're looking at the wrong boolean :) initOnce has its own flag, which it uses for its' double-checked locking. Variable `initialized` just happens to be a boolean.

So, what happens is:
- thread reads (atomically, but perhaps out of order) `initialized`, not using any fences.
- it **may** read `false` incorrectly, and consequently call `initOnce`. But it will never read `true` incorrectly.
- if it read `false` and called `initOnce`, `initOnce` has to make sure it doesn't read it's own flag out of order w.r.t. writes.

So Andrei's check **looks** redundant w.r.t. to `initOnce`, but it isn't, as it's guaranteed cheap.